sound/soc/soc-dapm.c

```diff
@@ -4464,9 +4464,27 @@ void snd_soc_dapm_connect_dai_link_widgets(struct snd_soc_card *card)
 			for_each_rtd_codec_dais(rtd, i, codec_dai)
 				dapm_connect_dai_pair(card, rtd, codec_dai,
 						      asoc_rtd_to_cpu(rtd, i));
+		} else if (rtd->dai_link->num_codecs > rtd->dai_link->num_cpus &&
+			   rtd->dai_link->codec_ch_maps) {
+			int cpu_id;
+
+			for_each_rtd_codec_dais(rtd, i, codec_dai) {
+				cpu_id = rtd->dai_link->codec_ch_maps[i].connected_cpu_id;
+				if (cpu_id >= rtd->dai_link->num_cpus) {
+					dev_err(card->dev,
+						"cpu_id %d is out of range\n", cpu_id);
+					continue;
+				}
+				dapm_connect_dai_pair(card, rtd, codec_dai,
+						      asoc_rtd_to_cpu(rtd, cpu_id));
+			}
 		} else {
 			dev_err(card->dev,
-				"N cpus to M codecs link is not supported yet\n");
+				"N cpus to M codecs link is not supported: ");
+			if (!rtd->dai_link->codec_ch_maps)
+				dev_err(card->dev, "no mapping func provided\n");
+			else
+				dev_err(card->dev, "codec number < cpu number\n");
```

Comment:
codec number %d < cpu number %d\n", rtd->dai_link->num_codecs, rtd->dai_link->num_cpus,

Reply:
is this if/else in the right block? Shouldnt it be in the else if{} block above?

Reply:
> is this if/else in the right block? Shouldnt it be in the else if{} block above?

Yes, they should be in the else{} block since they are invalid cases. The if/else here is to distinguish the reasons.

```
        else {
                        dev_err(card->dev,
                                "N cpus to M codecs link is not supported: ");
                        if (!rtd->dai_link->codec_ch_maps)
                                dev_err(card->dev, "no mapping func provided\n");
                        else
                                dev_err(card->dev, "codec number %d < cpu number %d\n",
                                        rtd->dai_link->num_codecs, rtd->dai_link->num_cpus);
                }
```